cmd/vfkit/main.go

```diff
@@ -264,3 +278,86 @@ func startIgnitionProvisionerServer(ignitionReader io.Reader, ignitionSocketPath
 	log.Debugf("ignition socket: %s", ignitionSocketPath)
 	return srv.Serve(listener)
 }
+
+func generateCloudInitImage(files []string) (string, error) {
+	if len(files) == 0 {
+		return "", nil
+	}
+
+	var userData, metaData string
+	for _, path := range files {
+		if path == "" {
+			continue
+		}
+		file, err := os.Open(path)
+		if err != nil {
+			return "", err
+		}
+		defer file.Close()
+
+		data, err := io.ReadAll(file)
+		if err != nil {
+			return "", err
+		}
+
+		// based on the file name we save its content on the corresponding var
+		filename := filepath.Base(path)
+		switch filename {
+		case "user-data":
+			userData = string(data)
+		case "meta-data":
+			metaData = string(data)
+		}
+	}
+
+	return createCloudInitISO(metaData, userData)
+}
+
+func createCloudInitISO(metaData, userData string) (string, error) {
+	writer, err := iso9660.NewWriter()
+	if err != nil {
+		return "", errors.Wrap(err, fmt.Sprintf("failed to create writer: %s", err))
+	}
+
+	defer writer.Cleanup()
+
+	userDataReader := strings.NewReader(userData)
+	err = writer.AddFile(userDataReader, "user-data")
+	if err != nil {
+		return "", errors.Wrap(err, fmt.Sprintf("failed to add file: %s", err))
+	}
+
+	metaDataReader := strings.NewReader(metaData)
+	err = writer.AddFile(metaDataReader, "meta-data")
+	if err != nil {
+		return "", errors.Wrap(err, fmt.Sprintf("failed to add file: %s", err))
+	}
+
+	cacheDir, err := os.UserCacheDir()
+	if err != nil {
+		return "", errors.Wrap(err, "unable to fetch user cache dir")
+	}
+
+	isoFilePath := filepath.Join(cacheDir, "vfkit-cloudinit.iso")
```

Comment:
is this file deleted after vfkit exits?  If not, I wonder if that filename should be somehow different per run? I dont know the origin of cachedir, so if the runs are namespaced, then you are probably cool.  But I was thinking of a scneario where two vms are run, both with cloud-init, would they overwrite eachother?

Reply:
yes, updated. It was fixed on that specific path bc of some specific script I was using for testing 👍 
but yes, there is no reason to keep it fixed, updated to a random filename